README.md
# multi_block
유해 사이트가 100만개이라고 가정을 하고 사용자가 접속하려는 사이트의 Host 이름이 100만개 리스트 안에 들어 갔을 때 패킷을 차단하는 프로그램을 작성 <br />

# Usage
./multiblock <br />
코드 내부에 iptables 와 netfilter_queue 명령어를 저장해뒀으니 그냥 쓰시면 됩니다.

# 코드 설명
python 파일로 일단 백만개의 url들을 각 .com .kr등등의 파일들로 나누고 dot_list라는 폴더에 저장한다. <br />
file pointer를 오픈시켜서 <br />

`
cat /bin/cat 경로/dot_list/url에 붙는 단어 | grep "필터링 될 url"
`
<br />의 방식으로 검색하였다. <br />
전역 변수 flag을 이용하여서 해당 Host 들을 필터링하였다.

# 보안할 점
url에 접속할 때 정확하게 url 입력해줘야한다.
예를 들어 clock.zone 이라 안하고 clock.zone/ 이라해도 그대로 뚫리니 문제.


dot_list/news
newstrend.news
elbalad.news
ura.news
lightnepal.news
apple.news
saat24.news
aftershock.news
socialtrend.news
kol7sry.news
baghdadtoday.news
nablus.news
bctv.news
shkodra.news
manoto.news
malaysiable.news
kabar.news
informator.news
postcard.news
favorites.news
enter.news
dazzling.news
happytify.news
tuko.news
monitorbd.news
chaskatimes.news
amdn.news
neuste.news
filo.news
vishwavani.news
begemot.news
watchers.news
dnn.news
kentlive.news
withinnigeria.news
avax.news
rubikon.news
textbook.news
pudhari.news
hania.news
cnread.news
frank.news
dandelion.news
bonnier.news
texts.news
tempo24.news
alsjl.news
volga.news
aum.news
susanin.news
reality.news
ongoma.news
needtoknow.news
kadenya.news
iranenergy.news
essexlive.news
nct.news
babe.news
ultimora.news
premieredate.news
manatelangana.news
knigi.news
ramallah.news
lawbooks.news
metro.news
zer.news
myanmarbreaking.news
privacypolicy.news
moetron.news
handball-world.news
orer.news
rewire.news
atlantea.news
socialpost.news
abomus.news
inside.news
uktech.news
celebz.news
neaselida.news
positive.news
imperium.news
bambuca.news
nigeriana.news
pattayaone.news
styl.news
arzuw.news
chrometv.news
rus24.news
censored.news
com-story.news
xn--l3c2b6a.news
chill.news
syria.news
khmernews.news
belltower.news
pakkatelugu.news
gamebet.news
digitaleng.news
lagazette.news
polizei.news
leopolis.news
bareta.news
crypto-currency.news
pleno.news
farandu.news
theliberator.news
pointnet.news
mixed.news
retty.news
thenewkhalij.news
anonymous-post.news
purboposhchimbd.news
anycoin.news
mos.news
dolphinpost.news
ideology.news
pornsites.news
adya.news
politicalmayhem.news
nicovideo.news
tsl.news
tamtam.news
kronos.news
article14.news
kairos.news
afghanistanonline.news
alotinmoi.news
diggers.news
amapapa.news
pendar.news
cyclingindustry.news
fromua.news
technocracy.news
stopeuro.news
alnabaa.news
amjilt.news
tvfb.news
huba.news
ultra.news
sharqkw.news
scoop24.news
negaam.news
uzreport.news
entreprise.news
chinatt.news
cryptocurrencyinvesting.news
allcryptocurrencies.news
iqraanews.news
binoculars.news
noodleremover.news
sunaulonepal.news
worldtime.news
bouncy.news
rutracker.news
therival.news
ildubbio.news
elrdar.news
pbd.news
ilia.news
tmb.news
survivethis.news
crank11.news
turningpoint.news
spotlights.news
upgo.news
direct.news
com-shows.news
tamilcinema.news
freedoms.news
nirvana.news
sheger.news
ajell.news
startupvalley.news
digital-life.news
padide.news
mozaika.news
bit.news
shootings.news
south.news
presseshop.news
icitizen.news
najm.news
russland.news
trucks.news
hyundai.news
kolo.news
ir24.news
kraichgau.news
dafuq.news
finbook.news
rodina.news
ivasi.news
noozup.news
khmerapsara.news
geopolitika.news
shocked.news
spesial.news
ilbitcoin.news
observer.news
trive.news
pause-cafe.news
wam-times.news
foot365.news
please.news
pushme.news
cryptocoin.news
citytoday.news
latestnewstoday.news
kafetogel.news
chretien.news
energy-storage.news
sumer.news
poraqui.news
dzvin.news
ua.news
sushantsinha.news
topaddis.news
dnepr.news
lagenda.news
ytfb.news
saudi50.news
qiqu.news
fsspx.news
saffron.news
smarttravel.news
cactus.news
cloudspace.news
coinmarket.news
mubasher24.news
airtable.news
androidtv.news
inkrculture.news
latest-hour.news
searchword.news
torresette.news
volgodonsk.news
dailynation.news
lurlo.news
genxsentinel.news
cnl.news
talkingsumo.news
dor.news
gadgets7.news
jellyshare.news
leg.news
theshblsh.news
radio1.news
guerrilla.news
fullmeasure.news
trump.news
triton.news
infokampus.news
crash-aerien.news
telanganatoday.news
thefappening.news
navigator.news
hvost.news
ppc.news
ardina.news
kaitori.news
misto.news
mychinese.news
elise.news
coinlaundry.news
proini.news
indiaspeaks.news
thegazette.news
oltrelalinea.news
tagmag.news
dacha.news
haikara.news
landbanking.news
prothombarta.news
thehook.news
4view.news
persian.news
biketoday.news
xaricdetehsil.news
observator.news
baghdad24.news
bit-economy.news
inlight.news
globalniportal.news
ignat.news
buzzy.news
1024.news
ikman.news
gmo.news
alertme.news
prism.news
fxcm.news
nabat.news
vivere.news
craigmedred.news
petsite.news
l34.news
revit.news
jagbharat.news
southernafrican.news
dailyherald.news
ytn.news
av18.news
cambodiadaily.news
cannatech.news
bacana.news
phim.news
mwebantu.news
medic.news
masto.news
marsad.news
alex24.news
geneonline.news
sahba.news
clashroyale.news
taranis.news
uglegorsk.news
doisong.news
xvideos.news
dailyreckoning.news
dojo.news
truedaily.news
com-jobs.news
duda.news
punjab24.news
fbs.news
apro.news
bomba.news
musical.news
dinajpur.news
rassd.news
clearpolitics.news
titbits.news
utro.news
irkutsk.news
xnxx.news
ntdaily.news
alfallahalyoum.news
action.news
khmershare.news
ifoodie.news
24-24.news
jobalert.news
otb.news
ternopil.news
mediaumat.news
tnpsc.news
dokan.news
g0v.news
cotedivoire.news
voice-yemen.news
happy88.news
seriale.news
suspilne.news
analytics.news
kamal.news
intim.news
phapluat.news
skener.news
khabran.news
eastnashville.news
danesh.news
kontho.news
estilofitness.news
socialist.news
wodoh.news
elclick.news
studionord.news
moq3k.news
recap-sport.news
magnis.news


dot_split.py
import os

# I used domain tail(?) such as .com .kr .jp .fr etc
f = open('top-1m.csv')
directory = './dot_list/'
char_set = []
dot = ''
count_dot = 0
put_url = {}
count = 0
os.system('mkdir -p dot_list')

while True:
	line = f.readline()
	if not line:
		break
	count = count + 1
	line = line.split(',')[1] # removing index
	char_set = line.split('.')
	dot = char_set[len(char_set)-1]
	dot = dot[:-1] # erasing \n

	#print char_set
	#print dot # .com / .kr / etc
	
	# dictionary ex) .com : www.naver.com
	if (dot in put_url.keys()): # .keys() helps to print out all the keys in put_url dict.
		for key in put_url:
			if (key == dot):
				f2 = open(directory + dot,'a+') # append to the files	
				f2.write(line)
				#	put_url[dot].append(line[:-1])
				f2.close()
	else:
		put_url[dot] = []
		f2 = open(directory + dot,'w')
		f2.write(line)
		f2.close()
		#put_url[dot].append(line[:-1])
	
	if (count % 10000 == 0):
		print count
		
"""
print 'finished ' + put_url
for key,value in put_url:	# value is list!
	print key,value
	f2 = open(directory + key + '.txt', 'w')
	for i in range(len(value)):
		f2.write(value[i]+'\n')
	f2.close()


	 """
f.close()


netfilter_block.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <netinet/in.h>
#include <linux/types.h>
#include <linux/netfilter.h>		/* for NF_ACCEPT */
#include <errno.h>
#include <string.h>

#include <netinet/ip.h>
// #include <netinet/ether.h>
#include <netinet/tcp.h>

#include <libnetfilter_queue/libnetfilter_queue.h>

#define IPTYPE 8
#define TCPTYPE 6

typedef u_int tcp_seq;

// Declare http_method & target (argv[1])
u_int8_t *target;
u_int8_t target_len;
u_int32_t id;
u_int8_t flag; // To identify ACCEPT, DROP


void iptables_F() {
	system("iptables -F");
}

void dump(unsigned char* buf, int size) {
    int i;
    for (i = 0; i < size; i++) {
        if (i % 16 == 0)
            printf("\n");
        printf("%02x ", buf[i]);
    }
	puts("");
}

// ---------------------------------------------------------------
static u_int32_t print_pkt (struct nfq_data *tb) {
	struct nfqnl_msg_packet_hdr *ph;
	struct nfqnl_msg_packet_hw *hwph;
	u_int32_t mark,ifi;
	int ret;
	unsigned char *data; // data for nfq_get_payload

	struct ip* ipHdr;
	struct tcphdr* tcpHdr;
	char * tcp_data_area;
	int ip_hdr_size;
	int tcp_hdr_size;
	int tcp_data_len;
	char * search_host;

	flag = 1;
	
	ph = nfq_get_msg_packet_hdr(tb);
	if(ph) id = ntohl(ph->packet_id); // id

	hwph = nfq_get_packet_hw(tb);
	if (hwph) {
		int i, hlen = ntohs(hwph->hw_addrlen);
	}

	mark = nfq_get_nfmark(tb);
	ifi = nfq_get_indev(tb);
	ifi = nfq_get_outdev(tb);
	ifi = nfq_get_physindev(tb);
	ifi = nfq_get_physoutdev(tb);
	ret = nfq_get_payload(tb, &data); // IP header의 시작 위치를 알아 낸다	
	if (ret >= 0) {
		printf("payload_len=%d ", ret);
		//dump(data, ret);
	} // finding IP header
	// Until this part, same as nfqtest!!!!!!!!
	int i =0;
	int size = ret;

	ipHdr = (struct ip *)data; //right
	ip_hdr_size = ipHdr->ip_hl * 4; //only ip header length
	//printf("[*] ipHdr size : %d\n", ip_hdr_size); // always 20 bytes
	//printf("[DUMP]ipHdr-----------");
	//dump((char *)ipHdr, ip_hdr_size);
	//printf("[ipHdr DUMP Finished]\n");

	// Now, let's find TCP Header & Data area! DUDE!!
	tcpHdr = (struct tcphdr *) (data + ip_hdr_size); // located at +20 
	tcp_hdr_size = (tcpHdr-> th_off * 4); // tcp header size
	tcp_data_area = data + ip_hdr_size + tcp_hdr_size;
	tcp_data_len = ret - ip_hdr_size - tcp_hdr_size; // ret is total length 

	if(ipHdr->ip_p == TCPTYPE && tcp_hdr_size > 0) {
		//printf("[*] tcpHdr size : %d\n",tcp_hdr_size);
		//printf("[DUMP]tcpHdr-----------");
		//dump((char *) tcpHdr, tcp_hdr_size);
		//printf("[tcpHdr DUMP Finished]\n");
		
		//printf("[*] tcp_data size : %d\n", tcp_data_len);
		//printf("[DUMP]tcp data area-----");
		//dump((char *) tcp_data_area, tcp_data_len);
		//printf("[tcp data area DUMP Finished]\n");
		//printf("%s\n", tcp_data_area);
		search_host = strstr(tcp_data_area, "Host: ");
		if(search_host != NULL) {
			// seg fault if it just print it!
			printf("****************ATTENTION***************\n");
			char *tmp;
			tmp = (char*)malloc(sizeof(char) * (6 + target_len)); // 6 for "Host: "
			strncpy(tmp, search_host, sizeof(char) * (6 + target_len));
			printf("%s\n", tmp);
			if (!strcmp(target,tmp+6)) {
				printf("[+] Same with the target DUDE!!!\n");
				flag = 0;
			}
			free(tmp);
		}
			
	}

	return id;
}


static int cb(struct nfq_q_handle *qh, struct nfgenmsg *nfmsg,
	      struct nfq_data *nfa, void *data)
{
	u_int32_t id = print_pkt(nfa);
	printf("[*] entering callback\n");
	return nfq_set_verdict(qh, id, flag, 0, NULL); // flag: 1 ACCEPT, 0 DROP
}

int main(int argc, char *argv[]) {
    struct nfq_handle *h;			// struct nfnl_handle * 	nfnlh
									// struct nfq_q_handle * 	qh_list
									// struct nfnl_subsys_handle * 	nfnlssh
    struct nfq_q_handle *qh;
    struct nfnl_handle *nh;
    int fd;
    int rv;
    char buf[4096] __attribute__ ((aligned));

	if (argc != 2) {
		puts("You need argv[1]!");
		exit(1);
	}
	
	target = argv[1];
	target_len = strlen(target);

	// Initializing the iptables configuration
	system("iptables -F"); //iptables_F
	system("iptables -A OUTPUT -j NFQUEUE --queue-num 0");
	system("iptables -A INPUT -j NFQUEUE --queue-num 0");

    printf("opening library handle\n");
    h = nfq_open(); // handler of a netfilter queue
    
	if (!h) {
        fprintf(stderr, "error during nfq_open()\n"); iptables_F();
        exit(1);
    }

    printf("unbinding existing nf_queue handler for AF_INET (if any)\n");
    if (nfq_unbind_pf(h, AF_INET) < 0) {
        fprintf(stderr, "error during nfq_unbind_pf()\n"); iptables_F();
        exit(1);
    }

    printf("binding nfnetlink_queue as nf_queue handler for AF_INET\n");
    if (nfq_bind_pf(h, AF_INET) < 0) {
        fprintf(stderr, "error during nfq_bind_pf()\n"); iptables_F();
        exit(1);
    }

    printf("binding this socket to queue '0'\n");
    qh = nfq_create_queue(h,  0, &cb, NULL);
    if (!qh) {
        fprintf(stderr, "error during nfq_create_queue()\n"); iptables_F();
        exit(1);
    }

    printf("setting copy_packet mode\n");
    if (nfq_set_mode(qh, NFQNL_COPY_PACKET, 0xffff) < 0) {
        fprintf(stderr, "can't set packet_copy mode\n"); iptables_F();
        exit(1);
    }

    fd = nfq_fd(h);	// usage: nfq_fd(handler)

    for (;;) {
        if ((rv = recv(fd, buf, sizeof(buf), 0)) >= 0) {

            printf("[+] pkt received\n");
            nfq_handle_packet(h, buf, rv);
            continue;
        }
        if (rv < 0 && errno == ENOBUFS) {
            printf("[-] losing packets!\n");
            continue;
        }
        perror("[-] recv failed");
        break;
    }

    printf("unbinding from queue 0\n");
    nfq_destroy_queue(qh);

#ifdef INSANE
    /* normally, applications SHOULD NOT issue this command, since
     * it detaches other programs/sockets from AF_INET, too ! */
    printf("unbinding from AF_INET\n");
    nfq_unbind_pf(h, AF_INET);
#endif

    printf("closing library handle\n");
    nfq_close(h);
	iptables_F();
    exit(0);

}


multi_block.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <netinet/in.h>
#include <linux/types.h>
#include <linux/netfilter.h>		/* for NF_ACCEPT */
#include <errno.h>
#include <string.h>

#include <netinet/ip.h>
// #include <netinet/ether.h>
#include <netinet/tcp.h>

#include <libnetfilter_queue/libnetfilter_queue.h>

#define IPTYPE 8
#define TCPTYPE 6

typedef u_int tcp_seq;

// Declare http_method & target (argv[1])
u_int8_t *target;
u_int8_t target_len;
u_int32_t id;
u_int8_t flag; // To identify ACCEPT, DROP


void iptables_F() {
	system("iptables -F");
}

void dump(unsigned char* buf, int size) {
    int i;
    for (i = 0; i < size; i++) {
        if (i % 16 == 0)
            printf("\n");
        printf("%02x ", buf[i]);
    }
	puts("");
}
int search_CRLF(char* search) { // in order to find target length ; before 0d 0a
	int _len = 0;
	
	for(; *search != '\0'; search++) {
    	/* Seach for a newline */
    	if(*search == '\n') {
      	  printf("\nnewline Found\n");
		  _len++;
   		}
	   	/* Search for a CRLF */
   		else if(*search == '\r' && *(search + 1) == '\n') {
   	    	printf("\nCRLF Found\n");
   	    	search++; /* search will be incremented twice (bit hacky?) */
			//search = *tmp; // getting back!
			return _len;
	  	}
		_len++;
	}
}

int file_open_find(char* dot, char* tmp) {
	char filename[300];
	char file_pwd[1000];
	int dot_len = strlen(dot);
	FILE *file_pointer;
	char cmd_buffer[400];
	char find_exist[100];
	
	FILE *fp;
	
	fp = popen("/bin/pwd","r");
	if (fp == NULL) {
		printf("I cannot find pwd!");
		exit(1);
	}
	while(fgets(file_pwd,sizeof(file_pwd)-2,fp) != NULL) {
		printf("[+] fgets: %s",file_pwd);
	}
	pclose(fp);
	file_pwd[strlen(file_pwd)-1] = '\0'; // '\0' instead of '\n' to use strcat!
	strcat(file_pwd,"/dot_list/");
	printf("[+] After strcat with /dot_list : %s\n",file_pwd);
	printf("file_pwd_length: %d",strlen(file_pwd));
	/*for (int i=0; i<dot_len; i++){
		printf("\ndot: %c",dot[i]);
		file_pwd[i+strlen(file_pwd)] = dot[i];
	}*/
	printf("\n");
	strcat(file_pwd,dot); // /home/rictr0y/gilgil/multi_block/dot_list/com
	printf("PLZ!!!!!!!!!!!!!: %s\n",file_pwd); // but segfault
	
	snprintf(cmd_buffer,100, "/bin/cat %s | grep \"%s\"",file_pwd,&tmp[6]); // &tmp[6] for after Host :
	printf("[+] cmd_buffer: %s\n",cmd_buffer);
	// 	filename finished
	// file_pointer=fopen(file_pwd, "r"); // file opened
	printf("cmd_buffer is on the system function!!!!!!!!\n");
	
	fp = popen(cmd_buffer,"r");
	if (fp == NULL) {
		printf("I cannot find cat!");
		exit(1);
	}
	while(fgets(find_exist,sizeof(find_exist)-2,fp) != NULL) {
		printf("[+] file exist! bro! : %s\n", find_exist);
		return 1;
	}
	return 0;

}

// ---------------------------------------------------------------
static u_int32_t print_pkt (struct nfq_data *tb) {
	struct nfqnl_msg_packet_hdr *ph;
	struct nfqnl_msg_packet_hw *hwph;
	u_int32_t mark,ifi;
	int ret;
	unsigned char *data; // data for nfq_get_payload

	struct ip* ipHdr;
	struct tcphdr* tcpHdr;
	char * tcp_data_area;
	int ip_hdr_size;
	int tcp_hdr_size;
	int tcp_data_len;
	char * search_host;

	flag = 1;
	
	ph = nfq_get_msg_packet_hdr(tb);
	if(ph) id = ntohl(ph->packet_id); // id

	hwph = nfq_get_packet_hw(tb);
	if (hwph) {
		int i, hlen = ntohs(hwph->hw_addrlen);
	}

	mark = nfq_get_nfmark(tb);
	ifi = nfq_get_indev(tb);
	ifi = nfq_get_outdev(tb);
	ifi = nfq_get_physindev(tb);
	ifi = nfq_get_physoutdev(tb);
	ret = nfq_get_payload(tb, &data); // IP header의 시작 위치를 알아 낸다	
	if (ret >= 0) {
		printf("payload_len=%d ", ret);
		//dump(data, ret);
	} // finding IP header
	// Until this part, same as nfqtest!!!!!!!!
	int i =0;
	int size = ret;

	ipHdr = (struct ip *)data; //right
	ip_hdr_size = ipHdr->ip_hl * 4; //only ip header length
	//printf("[*] ipHdr size : %d\n", ip_hdr_size); // always 20 bytes
	//printf("[DUMP]ipHdr-----------");
	//dump((char *)ipHdr, ip_hdr_size);
	//printf("[ipHdr DUMP Finished]\n");

	// Now, let's find TCP Header & Data area! DUDE!!
	tcpHdr = (struct tcphdr *) (data + ip_hdr_size); // located at +20 
	tcp_hdr_size = (tcpHdr-> th_off * 4); // tcp header size
	tcp_data_area = data + ip_hdr_size + tcp_hdr_size;
	tcp_data_len = ret - ip_hdr_size - tcp_hdr_size; // ret is total length 
	
	char hi[] = "/home/rictr0y/gilgil/multi_block/dot_list/";
	FILE *file_pointer;
	char *tmp;

	if(ipHdr->ip_p == TCPTYPE && tcp_hdr_size > 0) {
		//printf("[*] tcpHdr size : %d\n",tcp_hdr_size);
		//printf("[DUMP]tcpHdr-----------");
		//dump((char *) tcpHdr, tcp_hdr_size);
		//printf("[tcpHdr DUMP Finished]\n");
		
		//printf("[*] tcp_data size : %d\n", tcp_data_len);
		//printf("[DUMP]tcp data area-----");
		//dump((char *) tcp_data_area, tcp_data_len);
		//printf("[tcp data area DUMP Finished]\n");
		//printf("%s\n", tcp_data_area);
		search_host = strstr(tcp_data_area, "Host: "); // finding Host: pointer
		if(search_host != NULL) {
			// seg fault if it just print it!
			printf("\n****************ATTENTION***************");
			printf("\n%s\n",search_host);
			target_len = search_CRLF(search_host+6); // pointer remained & returns target_len
			printf("target_len: %d",target_len);
			search_host = strstr(tcp_data_area, "Host: "); // finding Host: again!
			tmp = (char*)malloc(sizeof(char) * (6 + target_len)); // 6 for "Host: " and 1 for \x00
			strncpy(tmp, search_host, sizeof(char) * (6 + target_len)); // ex) tmp -> Host: www.naver.com
			printf("\n%s\n", tmp);
			// we have to find whether tmp is in the list of 1m-top.csv
			//strrchr 문자열의 끝에서부터 역순으로 검색해서 문자를 찾았으면 해당 문자로 시작하는 문자열의 포인터를 반환, 문자가 없으면 NULL을 반환
			char *finded_dot = strrchr(tmp, '.');
			printf("%s\n", finded_dot);
				
			if (file_open_find((finded_dot+1),tmp)) { // not .com, only com! | and find www.naver.com
				printf("[+] Same with the target DUDE!!!\n");
				flag = 0;
			}	
			free(tmp);
		}	
	}
	return id;
}


static int cb(struct nfq_q_handle *qh, struct nfgenmsg *nfmsg,
	      struct nfq_data *nfa, void *data)
{
	u_int32_t id = print_pkt(nfa);
	printf("[*] entering callback\n");
	return nfq_set_verdict(qh, id, flag, 0, NULL); // flag: 1 ACCEPT, 0 DROP
}

int main(int argc, char *argv[]) {
    struct nfq_handle *h;			// struct nfnl_handle * 	nfnlh
									// struct nfq_q_handle * 	qh_list
									// struct nfnl_subsys_handle * 	nfnlssh
    struct nfq_q_handle *qh;
    struct nfnl_handle *nh;
    int fd;
    int rv;
    char buf[4096] __attribute__ ((aligned));
	
	// I will find the last dot from the url!!
	//char *finded_dot = strrchr(target, '.');
	//printf("%s\n", finded_dot);

		
	// Initializing the iptables configuration
	system("iptables -F"); //iptables_F
	system("iptables -A OUTPUT -j NFQUEUE --queue-num 0");
	system("iptables -A INPUT -j NFQUEUE --queue-num 0");

    printf("opening library handle\n");
    h = nfq_open(); // handler of a netfilter queue
    
	if (!h) {
        fprintf(stderr, "error during nfq_open()\n"); iptables_F();
        exit(1);
    }

    printf("unbinding existing nf_queue handler for AF_INET (if any)\n");
    if (nfq_unbind_pf(h, AF_INET) < 0) {
        fprintf(stderr, "error during nfq_unbind_pf()\n"); iptables_F();
        exit(1);
    }

    printf("binding nfnetlink_queue as nf_queue handler for AF_INET\n");
    if (nfq_bind_pf(h, AF_INET) < 0) {
        fprintf(stderr, "error during nfq_bind_pf()\n"); iptables_F();
        exit(1);
    }

    printf("binding this socket to queue '0'\n");
    qh = nfq_create_queue(h,  0, &cb, NULL);
    if (!qh) {
        fprintf(stderr, "error during nfq_create_queue()\n"); iptables_F();
        exit(1);
    }

    printf("setting copy_packet mode\n");
    if (nfq_set_mode(qh, NFQNL_COPY_PACKET, 0xffff) < 0) {
        fprintf(stderr, "can't set packet_copy mode\n"); iptables_F();
        exit(1);
    }

    fd = nfq_fd(h);	// usage: nfq_fd(handler)

    for (;;) {
        if ((rv = recv(fd, buf, sizeof(buf), 0)) >= 0) {

            printf("[+] pkt received\n");
            nfq_handle_packet(h, buf, rv);
            continue;
        }
        if (rv < 0 && errno == ENOBUFS) {
            printf("[-] losing packets!\n");
            continue;
        }
        perror("[-] recv failed");
        break;
    }

    printf("unbinding from queue 0\n");
    nfq_destroy_queue(qh);

#ifdef INSANE
    /* normally, applications SHOULD NOT issue this command, since
     * it detaches other programs/sockets from AF_INET, too ! */
    printf("unbinding from AF_INET\n");
    nfq_unbind_pf(h, AF_INET);
#endif

    printf("closing library handle\n");
    nfq_close(h);
	iptables_F();
    exit(0);

}
